### ISSUE TITLE
PHP 7.3 segfaults

### ISSUE THREAD

**Issue by paragonie-scott** [jedisct1/libsodium-php#186]
```php
<?php

function foo()
{
    throw new SodiumException('test');
}
foo();

echo 'OK', PHP_EOL;
```

Expected result:

```
PHP Fatal error:  Uncaught SodiumException: test in /home/scott/sodium_compat/segfault.php:5
Stack trace:
#0 /home/scott/sodium_compat/segfault.php(7): foo()
#1 {main}
  thrown in /home/scott/sodium_compat/segfault.php on line 5
```

Actual result:

```
Segmentation fault (core dumped)
```

**Comment by paragonie-scott** [comment]
If you have a `throw new SodiumException` by itself, no segfault. From within a function or method, it segfaults. This tells me it's corrupting the callstack somehow.

**Comment by paragonie-scott** [comment]
#187 adds this condition to the test script.

**Comment by paragonie-scott** [comment]
Also reported to the PHP team here: https://bugs.php.net/bug.php?id=77297

**Comment by jedisct1** [comment]
`sodium_exception_create_object()` tries to remove function arguments from the backtrace, and this is probably where things break.

But I'm not familiar enough with the PHP internals to understand that code.

And idea, @nikic ?

**Comment by paragonie-scott** [comment]
A recent comment on php.net points to a possible cause:

 [2018-12-14 12:10 UTC] cmb@php.net

> 
> It seems the problem is that zend_hash_clean() is called[1] on a
> hash table with refcount==2, which is not allowed:
> 
>     php: /mnt/c/Users/cmb/php-dev/php-src/Zend/zend_hash.c:1521: zend_hash_clean: Assertion `(zend_gc_refcount(&(ht)->gc) == 1) || ((ht)->u.flags & (1<<6))' failed.
>     Aborted (core dumped)
> 
> Also `Z_TYPE_P(frame) == IS_ARRAY` looks fishy; shouldn't that be
`Z_TYPE_P(args) == IS_ARRAY`?
> 
> [1] <https://github.com/php/php-src/blob/php-7.3.0/ext/sodium/libsodium.c#L390-L392>

**Comment by jedisct1** [comment]
Seems to be fixed by 736b7d1e26e0aee100b5eade39c65664ef9e8e9e

Thanks!
